manual/coding_standards.tex

```diff
@@ -0,0 +1,499 @@
+
+\subsection{File Names}
+Each class should be defined in a separate file whose name is the same as the class name. Use separate \inlinecode{.cpp} implementation files whenever possible to aid in incremental compilation. 
+
+The filenames of tests are composed by the filename of the object tested and the prefix \inlinecode{test_}.
+The filenames of \emph{fake} and \emph{mock} objects used in tests are composed by the prefixes \inlinecode{fake_} and \inlinecode{mock_}, respectively, and the filename of the object that is imitated.
+
+\subsection{Header files}
+All header files should be self-contained i.e. it is not dependent on following any other header when it is included. Nor should they include files that are not necessary for their use, i.e. headers only needed by the implementation. Implementation files should not include files only for the benefit of files they include.
+
+There are many header files that violate this currently.
+Each header must use \inlinecode{\#define} guards to prevent multiple inclusion.
+The symbol name of the \inlinecode{\#define} guards should be \inlinecode{NAMESPACE(s)_CLASSNAME_H}.
+
+\subsection{Includes}
+Header files should be included with the full path based on the \verb|src| directory.
+For example, the file \verb|qmcpack/src/QMCWaveFunctions/SPOSetBase.h| should be included as
+\begin{lstlisting}
+#include "QMCWaveFunctions/SPOSetBash.h"
+\end{lstlisting}
+Even if the included file is located in the same directory as the including file this rule should be obeyed. Header files from external projects and standard libraries should be includes using \inlinecode{<iostream>} convention, while headers that are part of the QMCPACK project should be included using the \verb|"our_header.h"| convention.
+
+For readability it is suggested to use the standard order of includes:
+\begin{enumerate}
+	\item related header
+	\item std C library headers
+	\item std C++ library headers
+	\item Other libraries' headers
+	\item QMCPACK headers
+\end{enumerate}
+
+In each section the included files should be sorted in alphabetical order.
+
+\section{Naming}
+The balance between description and ease of implementation should be balanced such that the code remains self documenting within a single terminal window.  If an extremely short variable name is used its scope must be shorter than $\sim 40$ lines. An exception is made for template parameters which must be in all CAPS.
+
+\subsection{Type and Class Names}
+
+Type and class names should start with a capital letter and have a capital letter for each new word.
+Underscores \inlinecode{_} are not allowed. 
+\subsection{Namespace Names}
+
+Namespace names should be one word, lowercase.
+
+\subsection{Variable Names}
+
+Variable names should not begin with a capital letter this is reserved for type and class names. Mixed caps or underscores to separate words are both permitable but only one convention should be used in a file.
```

Comment:
Strongly prefer names like this_variable_name rather than thisVariableName (applying to everything but class names).  Reasonable to set this as a rule for new contributions?

Reply:
I disagree, to me thisVariableName is nicer than this_variable_name

Reply:
Mixed feelings. VariableName is good. However, when there is abbreviations I tend to add underscore. For example, SPOSet is actually a bit hard to read. Once you know QMCPACK for a while, it is not a problem.